client/apifetch.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * Middlewares to be applied to the apiFetch call.
+ */
+const middlewareRegistry = [];
+
+/**
+ * Applies all middlewares and runs the request.
+ *
+ * @param  {Object}   options      Passed as the second param for window.fetch
+ * @param  {string}   options.path Request URL
+ * @param  {Function} apply        The middleware to apply next
+ * @return {Promise}               Request promise
+ */
+const run = async ( { path, ...options }, [ apply, ...middlewares ] ) => {
+	if ( ! apply ) {
+		const response = await window.fetch( path, options );
+
+		return response.json();
+	}
+
+	return apply(
+		{
+			path,
+			...options,
+		},
+		( nextOptions ) => run( nextOptions, middlewares )
+	);
+};
+
+/**
+ * Makes a request using window.fetch and registered middleware.
+ *
+ * @param  {Object}  options      Request options
+ * @param  {string}  options.path Request URL (required)
+ * @return {Promise}              Request promise
+ */
+const apiFetch = async ( options ) => run( options, middlewareRegistry );
+
+/**
+ * Appends a middleware to apiFetch
+ *
+ * @param  {Function} middleware Middleware function
+ */
+apiFetch.use = ( middleware ) => middlewareRegistry.push( middleware );
+
+/**
+ * Prefix the request URLs with '/wp-json'
+ */
+apiFetch.use( ( options, next ) => {
```

Comment:
If this middleware is hardcoded here, how would we go adding another one for a different env where the path is completely different?

Maybe I'm still not grasping the concept right, but trying to add another middleware to manipulate the path seems hacky at this point instead of flexible?